File: app/src/main/java/de/haukehinrichs/osmview/SearchInputListener.java
package de.haukehinrichs.osmview;

import android.os.AsyncTask;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.SearchView;

import com.google.android.gms.maps.model.LatLng;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.util.ArrayList;

/**
 * Created by hauke on 09/03/17.
 */

public class SearchInputListener implements SearchView.OnSuggestionListener, SearchView.OnQueryTextListener {

    private String mApiKey;
    private LatLng mLocation;
    private ArrayAdapter<String> mAdapter;
    private ArrayList<String> mSuggestions;
    private JSONArray mResultList;
    private MainActivity mContext;

    SearchInputListener(String apiKey, ListView suggestionView, MainActivity context) {
        mApiKey = apiKey;
        mSuggestions = new ArrayList<>();
        mAdapter = new ArrayAdapter<String>(suggestionView.getContext(), R.layout.suggestion_list_item, R.id.suggestion_item, mSuggestions);
        suggestionView.setAdapter(mAdapter);
        suggestionView.setOnItemClickListener(new SearchInputClickListener());
        mContext = context;
    }

    public void updateLocation(LatLng location) {
        mLocation = location;
    }

    @Override
    public boolean onSuggestionSelect(int position) {
        return false;
    }

    @Override
    public boolean onSuggestionClick(int position) {
        return false;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        String urlString = "https://maps.googleapis.com/maps/api/place/autocomplete/json?key=" + mApiKey
                + "&input=" + URLEncoder.encode(newText);
        if (mLocation != null) {
            urlString +=  "&radius=50&location=" + Double.toString(mLocation.latitude) + "," + Double.toString(mLocation.longitude);
        }
        Log.d("places", "searching for " + newText);
        new SearchSuggestionsTask().execute(new String[] {urlString});
        return true;
    }

    private void handleSuggestionsResult(JSONObject result) throws JSONException {
        mResultList = result.getJSONArray("predictions");
        mSuggestions.clear();
        for (int i = 0; i < mResultList.length(); i++) {
            JSONObject singleResult = mResultList.getJSONObject(i);
            String description = singleResult.getString("description");
            Log.d("places", "result: " + description);
            mSuggestions.add(description);
        }
        mAdapter.notifyDataSetChanged();
    }

    class SearchSuggestionsTask extends AsyncTask<String, Void, JSONObject> {

        @Override
        protected JSONObject doInBackground(String... params) {

        try {
            URL url = new URL(params[0]);
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            Log.d("places", "opening url " + params[0]);
            try {
                BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(urlConnection.getInputStream()));
                StringBuilder stringBuilder = new StringBuilder();
                String line;
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line).append("\n");
                }
                bufferedReader.close();
                JSONObject result = new JSONObject(stringBuilder.toString());
                Log.d("json url result", "status " + result.getString("status"));
                return result;
            } finally {
                urlConnection.disconnect();
            }
        } catch(Exception e) {
            Log.e("ERROR", e.getMessage(), e);
            return null;
        }
        }

        @Override
        protected void onPostExecute(JSONObject jsonObject) {
            try {
                handleSuggestionsResult(jsonObject);
            } catch(Exception e) {
                Log.e("ERROR", e.getMessage(), e);
            }
        }
    }

    class SearchResultTask extends SearchSuggestionsTask {
        @Override
        protected void onPostExecute(JSONObject jsonObject) {
            try {
                handlePlaceResult(jsonObject);
            } catch(Exception e) {
                Log.e("ERROR", e.getMessage(), e);
            }
        }
    }

    private void handlePlaceResult(JSONObject result) throws JSONException {
        JSONObject location = result.getJSONObject("result").getJSONObject("geometry").getJSONObject("location");
        double lat = location.getDouble("lat");
        double lng = location.getDouble("lng");
        LatLng latLng = new LatLng(lat, lng);
        mContext.addMarker(latLng, result.getJSONObject("result").getString("name"));
        mSuggestions.clear();
        mAdapter.notifyDataSetChanged();
    }


    class SearchInputClickListener implements AdapterView.OnItemClickListener {

        @Override
        public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
            try {
                JSONObject place = mResultList.getJSONObject(position);
                String placeId = place.getString("place_id");
                String urlString = "https://maps.googleapis.com/maps/api/place/details/json?key=" + mApiKey
                        + "&placeid=" + placeId;
                Log.d("places", "looking up " + placeId);
                new SearchResultTask().execute(new String[] {urlString});
            } catch (JSONException e) {
                Log.e("ERROR", e.getMessage(), e);
            }

        }
    }
}

File: app/src/main/java/de/haukehinrichs/osmview/SearchInputListener.kt
package de.haukehinrichs.osmview

import android.os.AsyncTask
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.SearchView

import com.google.android.gms.maps.model.LatLng

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.ArrayList

/**
 * Created by hauke on 09/03/17.
 */

class SearchInputListener internal constructor(private val mApiKey: String, suggestionView: ListView, private val mContext: MainActivity) : SearchView.OnSuggestionListener, SearchView.OnQueryTextListener {
    private var mLocation: LatLng? = null
    private val mAdapter: ArrayAdapter<String>
    private val mSuggestions: ArrayList<String>
    private var mResultList: JSONArray? = null

    init {
        mSuggestions = ArrayList<String>()
        mAdapter = ArrayAdapter(suggestionView.context, R.layout.suggestion_list_item, R.id.suggestion_item, mSuggestions)
        suggestionView.adapter = mAdapter
        suggestionView.onItemClickListener = SearchInputClickListener()
    }

    fun updateLocation(location: LatLng) {
        mLocation = location
    }

    override fun onSuggestionSelect(position: Int): Boolean {
        return false
    }

    override fun onSuggestionClick(position: Int): Boolean {
        return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    override fun onQueryTextChange(newText: String): Boolean {
        var urlString = "https://maps.googleapis.com/maps/api/place/autocomplete/json?key=" + mApiKey + "&input=" + URLEncoder.encode(newText)
        if (mLocation != null) {
            urlString += "&radius=50&location=" + java.lang.Double.toString(mLocation!!.latitude) + "," + java.lang.Double.toString(mLocation!!.longitude)
        }
        Log.d("places", "searching for " + newText)
        SearchSuggestionsTask().execute(*arrayOf(urlString))
        return true
    }

    @Throws(JSONException::class)
    private fun handleSuggestionsResult(result: JSONObject) {
        mResultList = result.getJSONArray("predictions")
        mSuggestions.clear()
        for (i in 0..mResultList!!.length() - 1) {
            val singleResult = mResultList!!.getJSONObject(i)
            val description = singleResult.getString("description")
            Log.d("places", "result: " + description)
            mSuggestions.add(description)
        }
        mAdapter.notifyDataSetChanged()
    }

    internal open inner class SearchSuggestionsTask : AsyncTask<String, Void, JSONObject>() {

        override fun doInBackground(vararg params: String): JSONObject? {

            try {
                val url = URL(params[0])
                val urlConnection = url.openConnection() as HttpURLConnection
                Log.d("places", "opening url " + params[0])
                try {
                    val reader = urlConnection.inputStream.buffered().reader()
                    val result = JSONObject(reader.readText())
                    reader.close()
                    Log.d("json url result", "status " + result.getString("status"))
                    return result
                } finally {
                    urlConnection.disconnect()
                }
            } catch (e: Exception) {
                Log.e("ERROR", e.message, e)
                return null
            }

        }

        override fun onPostExecute(jsonObject: JSONObject) {
            try {
                handleSuggestionsResult(jsonObject)
            } catch (e: Exception) {
                Log.e("ERROR", e.message, e)
            }

        }
    }

    internal inner class SearchResultTask : SearchSuggestionsTask() {
        override fun onPostExecute(jsonObject: JSONObject) {
            try {
                handlePlaceResult(jsonObject)
            } catch (e: Exception) {
                Log.e("ERROR", e.message, e)
            }

        }
    }

    @Throws(JSONException::class)
    private fun handlePlaceResult(result: JSONObject) {
        val location = result.getJSONObject("result").getJSONObject("geometry").getJSONObject("location")
        val lat = location.getDouble("lat")
        val lng = location.getDouble("lng")
        val latLng = LatLng(lat, lng)
        mContext.addMarker(latLng, result.getJSONObject("result").getString("name"))
        mSuggestions.clear()
        mAdapter.notifyDataSetChanged()
    }


    internal inner class SearchInputClickListener : AdapterView.OnItemClickListener {

        override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
            try {
                val place = mResultList!!.getJSONObject(position)
                val placeId = place.getString("place_id")
                val urlString = "https://maps.googleapis.com/maps/api/place/details/json?key=" + mApiKey + "&placeid=" + placeId
                Log.d("places", "looking up " + placeId)
                SearchResultTask().execute(*arrayOf(urlString))
            } catch (e: JSONException) {
                Log.e("ERROR", e.message, e)
            }

        }
    }
}
